File: example-android-project/app/src/main/java/io/github/gofaith/faithdroid/FViews/FWebView.java
package io.github.gofaith.faithdroid.FViews;

import android.net.Uri;
import android.os.Build;
import android.support.v4.view.ViewCompat;
import android.util.Log;
import android.view.View;
import android.view.ViewGroup;
import android.webkit.DownloadListener;
import android.webkit.WebResourceRequest;
import android.webkit.WebView;
import android.webkit.WebViewClient;

import org.json.JSONObject;
import org.json.JSONTokener;

import java.util.HashMap;
import java.util.Iterator;
import java.util.List;
import java.util.Map;

import faithdroid.Faithdroid;
import io.github.gofaith.faithdroid.UI.AttrGettable;
import io.github.gofaith.faithdroid.UI.AttrSettable;
import io.github.gofaith.faithdroid.UI.UIController;

public class FWebView extends FView implements AttrGettable, AttrSettable {
    public WebView v;

    public FWebView(UIController controller) {
        this.parentController=controller;
        v = new WebView(parentController.activity);
        v.setId(ViewCompat.generateViewId());
        view=v;
        v.getSettings().setDefaultTextEncodingName("UTF -8");
        parentController.onDestroyEvent.add(new Runnable() {
            @Override
            public void run() {
                if (v!=null) {
                    ViewGroup parent = (ViewGroup) v.getParent();
                    if (parent != null) {
                        parent.removeView(v);
                    }
                    v.stopLoading();
                    v.getSettings().setJavaScriptEnabled(false);
                    v.clearHistory();
                    v.clearView();
                    v.removeAllViews();
                    v.destroy();
                    v = null;
                }
                parentController.viewmap.remove(vID);
            }
        });
    }
    @Override
    public String getAttr(String attr) {
        String str = getUniversalAttr(attr);
        if (str != null) {
            return str;
        }
        switch (attr) {
        }
        return null;
    }

    @Override
    public void setAttr(String attr, final String value) {
        if (setUniversalAttr(attr, value)) {
            return;
        }
        switch (attr) {
            // -------------------------------------------------------------------
            case "Uri":
                v.loadUrl(value);
                break;
            case "Focusable":
                    v.setFocusable(value.equals("true"));
                break;
            case "SupportZoom":
                    v.getSettings().setSupportZoom(value.equals("true"));
                break;
            case "BuiltInZoomControls":
                v.getSettings().setBuiltInZoomControls(value.equals("true"));
                break;
            case "UseWideViewPort":
                v.getSettings().setUseWideViewPort(value.equals("true"));
                break;
            case "AllowContentAccess":
                v.getSettings().setAllowContentAccess(value.equals("true"));
                break;
            case "AllowFileAccess":
                v.getSettings().setAllowFileAccess(value.equals("true"));
                break;
            case "AllowFileAccessFromFileURLs":
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN) {
                    v.getSettings().setAllowFileAccessFromFileURLs(value.equals("true"));
                }
                break;
            case "AllowUniversalAccessFromFileURLs":
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN) {
                    v.getSettings().setAllowUniversalAccessFromFileURLs(value.equals("true"));
                }
                break;
            case "AppCacheEnabled":
                v.getSettings().setAppCacheEnabled(value.equals("true"));
                break;
            case "BlockNetworkImage":
                v.getSettings().setBlockNetworkImage(value.equals("true"));
                break;
            case "BlockNetworkLoads":
                v.getSettings().setBlockNetworkLoads(value.equals("true"));
                break;
            case "JavaScriptEnabled":
                v.getSettings().setJavaScriptEnabled(value.equals("true"));
                break;
            case "OffscreenPreRaster":
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
                    v.getSettings().setOffscreenPreRaster(value.equals("true"));
                }
                break;
            case "SaveFormData":
                v.getSettings().setSaveFormData(value.equals("true"));
                break;
            case "SupportMultipleWindows":
                v.getSettings().setSupportMultipleWindows(value.equals("true"));
                break;
            case "TextZoom":
                try {
                    v.getSettings().setTextZoom(Integer.parseInt(value));
                } catch (Exception e) {
                    e.printStackTrace();
                }
                break;
            case "UserAgentString":
                v.getSettings().setUserAgentString(value);
                break;
                // -------------------
            case "HandleUrl":
                MyWebViewClient wc=new MyWebViewClient();
                if (!value.equals("")) {
                    try {
                        wc.urlHandlersMap = new HashMap<>();
                        JSONObject object = (JSONObject) (new JSONTokener(value).nextValue());
                        Log.d(TAG, "setAttr: HandleUrl"+value);
                        for (Iterator<String> it = object.keys(); it.hasNext(); ) {
                            String key = it.next();
                            wc.urlHandlersMap.put(key, object.getString(key));
                        }
                    } catch (Exception e) {
                        e.printStackTrace();
                    }
                }
                v.setWebViewClient(wc);
                break;
            case "DownloadListener":
                v.setDownloadListener(new DownloadListener() {
                    @Override
                    public void onDownloadStart(String url, String userAgent, String contentDisposition, String mimetype, long contentLength) {
                        Faithdroid.triggerEventHandler(value, url);
                    }
                });
                break;
            case "LoadHtmlData":
                String a = value.replace("#", "%23");
                v.loadData(a, "text/html; charset=UTF-8", null);
                break;
        }
    }
    class MyWebViewClient extends WebViewClient{
        public HashMap<String,String> urlHandlersMap=null;
        @Override
        public boolean shouldOverrideUrlLoading(WebView view, WebResourceRequest request) {
            if (urlHandlersMap==null) {
                return super.shouldOverrideUrlLoading(view, request);
            }else {
                Log.d(TAG, "shouldOverrideUrlLoading: "+urlHandlersMap.size());
                if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.LOLLIPOP) {
                    String url = request.getUrl().toString();
                    if (urlHandlersMap.containsKey(removeQuery(url))) {
                        String fnId = urlHandlersMap.get(removeQuery(url));
                        String s= Faithdroid.triggerEventHandler(fnId, url);
                        if(s!=null&&s.equals("true"))
                            return true;
                    }else if (urlHandlersMap.containsKey(url)) {
                        String fnId = urlHandlersMap.get(url);
                        String s= Faithdroid.triggerEventHandler(fnId, url);
                        if(s!=null&&s.equals("true"))
                            return true;
                    }
                }
            }
            return false;
        }

        @Override
        public boolean shouldOverrideUrlLoading(WebView view, String url) {
            if (urlHandlersMap==null) {
                return super.shouldOverrideUrlLoading(view, url);
            }else {
                Log.d(TAG, "shouldOverrideUrlLoading2: "+urlHandlersMap.size());
                if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.LOLLIPOP) {
                    if (urlHandlersMap.containsKey(removeQuery(url))) {
                        String fnId = urlHandlersMap.get(removeQuery(url));
                        String s= Faithdroid.triggerEventHandler(fnId, url);
                        if(s!=null&&s.equals("true"))
                            return true;
                    }else if (urlHandlersMap.containsKey(url)) {
                        String fnId = urlHandlersMap.get(url);
                        String s= Faithdroid.triggerEventHandler(fnId, url);
                        if(s!=null&&s.equals("true"))
                            return true;
                    }
                }
            }
            return false;
        }
    }

    public static String removeQuery(String url) {
        int i = url.indexOf("?");
        if (i < 0) {
            return url;
        }
        return url.substring(0, i);
    }
}

File: android/app/src/main/java/io/github/stevenzack/faithdroidapp/FViews/FWebView.kt
package io.github.stevenzack.faithdroidapp.FViews

import android.content.ContentValues.TAG
import android.os.Build
import android.util.Log
import android.view.ViewGroup
import android.webkit.WebResourceRequest
import android.webkit.WebView
import android.webkit.WebViewClient
import faithdroid.Faithdroid
import io.github.stevenzack.faithdroidapp.UI.AttrGettable
import io.github.stevenzack.faithdroidapp.UI.AttrSettable
import io.github.stevenzack.faithdroidapp.UI.FView
import io.github.stevenzack.faithdroidapp.UI.UIController
import org.json.JSONObject
import org.json.JSONTokener


class FWebView(controller: UIController) : FView(), AttrGettable, AttrSettable {
    var v: WebView? = null

    init {
        this.parentController = controller
        v = WebView(parentController!!.activity)
        view = v
        v!!.settings.defaultTextEncodingName = "UTF -8"
        parentController!!.onDestoryEvent.add( {
            if (v != null) {
                val parent = v!!.parent as ViewGroup
                parent?.removeView(v)
                v!!.stopLoading()
                v!!.settings.javaScriptEnabled = false
                v!!.clearHistory()
                v!!.clearView()
                v!!.removeAllViews()
                v!!.destroy()
                v = null
            }
            parentController!!.viewmap.remove(vID)
        })
    }

    override fun getAttr(attr: String): String{
        val str = getUniversalAttr(attr)
        if (str != null) {
            return str
        }
        when (attr) {

        }
        return ""
    }

    override fun setAttr(attr: String, value: String) {
        if (setUniversalAttr(attr, value)) {
            return
        }
        when (attr) {
            // -------------------------------------------------------------------
            "Uri" -> v!!.loadUrl(value)
            "Focusable" -> v!!.isFocusable = value == "true"
            "SupportZoom" -> v!!.settings.setSupportZoom(value == "true")
            "BuiltInZoomControls" -> v!!.settings.builtInZoomControls = value == "true"
            "UseWideViewPort" -> v!!.settings.useWideViewPort = value == "true"
            "AllowContentAccess" -> v!!.settings.allowContentAccess = value == "true"
            "AllowFileAccess" -> v!!.settings.allowFileAccess = value == "true"
            "AllowFileAccessFromFileURLs" -> if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN) {
                v!!.settings.allowFileAccessFromFileURLs = value == "true"
            }
            "AllowUniversalAccessFromFileURLs" -> if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN) {
                v!!.settings.allowUniversalAccessFromFileURLs = value == "true"
            }
            "AppCacheEnabled" -> v!!.settings.setAppCacheEnabled(value == "true")
            "BlockNetworkImage" -> v!!.settings.blockNetworkImage = value == "true"
            "BlockNetworkLoads" -> v!!.settings.blockNetworkLoads = value == "true"
            "JavaScriptEnabled" -> v!!.settings.javaScriptEnabled = value == "true"
            "OffscreenPreRaster" -> if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
                v!!.settings.offscreenPreRaster = value == "true"
            }
            "SaveFormData" -> v!!.settings.saveFormData = value == "true"
            "SupportMultipleWindows" -> v!!.settings.setSupportMultipleWindows(value == "true")
            "TextZoom" -> try {
                v!!.settings.textZoom = Integer.parseInt(value)
            } catch (e: Exception) {
                e.printStackTrace()
            }

            "UserAgentString" -> v!!.settings.setUserAgentString(value)
            // -------------------
            "HandleUrl" -> {
                val wc = MyWebViewClient()
                if (value != "") {
                    try {
                        wc.urlHandlersMap = HashMap()
                        val `object` = JSONTokener(value).nextValue() as JSONObject
                        Log.d(TAG, "setAttr: HandleUrl$value")
                        val it = `object`.keys()
                        while (it.hasNext()) {
                            val key = it.next()
                            wc.urlHandlersMap!![key] = `object`.getString(key)
                        }
                    } catch (e: Exception) {
                        e.printStackTrace()
                    }

                }
                v!!.webViewClient = wc
            }
            "DownloadListener" -> v!!.setDownloadListener { url, userAgent, contentDisposition, mimetype, contentLength ->
                Faithdroid.triggerEventHandler(
                    value,
                    url
                )
            }
            "LoadHtmlData" -> {
                val a = value.replace("#", "%23")
                v!!.loadData(a, "text/html; charset=UTF-8", null)
            }
        }
    }

    internal inner class MyWebViewClient : WebViewClient() {
        var urlHandlersMap: HashMap<String, String>? = null
        override fun shouldOverrideUrlLoading(view: WebView, request: WebResourceRequest): Boolean {
            if (urlHandlersMap == null) {
                return super.shouldOverrideUrlLoading(view, request)
            } else {
                Log.d(TAG, "shouldOverrideUrlLoading: " + urlHandlersMap!!.size)
                if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.LOLLIPOP) {
                    val url = request.url.toString()
                    if (urlHandlersMap!!.containsKey(removeQuery(url))) {
                        val fnId = urlHandlersMap!![removeQuery(url)]
                        val s = Faithdroid.triggerEventHandler(fnId, url)
                        if (s != null && s == "true")
                            return true
                    } else if (urlHandlersMap!!.containsKey(url)) {
                        val fnId = urlHandlersMap!![url]
                        val s = Faithdroid.triggerEventHandler(fnId, url)
                        if (s != null && s == "true")
                            return true
                    }
                }
            }
            return false
        }

        override fun shouldOverrideUrlLoading(view: WebView, url: String): Boolean {
            if (urlHandlersMap == null) {
                return super.shouldOverrideUrlLoading(view, url)
            } else {
                Log.d(TAG, "shouldOverrideUrlLoading2: " + urlHandlersMap!!.size)
                if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.LOLLIPOP) {
                    if (urlHandlersMap!!.containsKey(removeQuery(url))) {
                        val fnId = urlHandlersMap!![removeQuery(url)]
                        val s = Faithdroid.triggerEventHandler(fnId, url)
                        if (s != null && s == "true")
                            return true
                    } else if (urlHandlersMap!!.containsKey(url)) {
                        val fnId = urlHandlersMap!![url]
                        val s = Faithdroid.triggerEventHandler(fnId, url)
                        if (s != null && s == "true")
                            return true
                    }
                }
            }
            return false
        }
    }

    companion object {

        fun removeQuery(url: String): String {
            val i = url.indexOf("?")
            return if (i < 0) {
                url
            } else url.substring(0, i)
        }
    }
}
